### ISSUE TITLE
GMRES segfault when preconditioner="NOSOLVER"

### ISSUE THREAD

**Issue by shwina** [NVIDIA/AMGX#28]
The GMRES solver produces a segfault when preconditioner="NOSOLVER":

```
[atrikut@node1144 examples]$ cat ../configs/core/GMRES.json
{
    "config_version": 2,
    "solver": {
        "preconditioner": {
            "scope": "amg",
            "solver": "NOSOLVER"
        },
        "use_scalar_norm": 1,
        "solver": "GMRES",
        "print_solve_stats": 1,
        "obtain_timings": 1,
        "monitor_residual": 1,
        "convergence": "RELATIVE_INI_CORE",
        "scope": "main",
        "tolerance": 1e-6,
        "norm": "L2"
    }
}
[atrikut@node1144 examples]$ ./amgx_capi -m matrix.mtx -c ../configs/core/GMRES.json

▽
  1 {
AMGX version 2.0.0.130-opensource
Built on May 17 2018, 05:29:30
Compiled with CUDA Runtime 8.0, using CUDA driver 9.0
Warning: No mode specified, using dDDI by default.
Reading data...
RHS vector was not found. Using RHS b=[1,…,1]^T
Solution vector was not found. Setting initial solution to x=[0,…,0]^T
Finished reading
           iter      Mem Usage (GB)       residual           rate
         --------------------------------------------------------------
            Ini                   0   3.464102e+00
              0                   0   1.845471e+00         0.5327
              1              0.0000   1.541877e+00         0.8355
              2              0.0000   1.374225e+00         0.8913
              3              0.0000   1.366903e+00         0.9947
              4              0.0000   1.040855e+00         0.7615
              5              0.0000   1.026638e+00         0.9863
              6              0.0000   8.614123e-01         0.8391
              7              0.0000   6.599583e-01         0.7661
              8              0.0000   6.596676e-01         0.9996
              9              0.0000   6.593714e-01         0.9996
             10              0.0000   6.331763e-01         0.9603
Caught signal 11 - SIGSEGV (segmentation violation)
 /home/atrikut/local/AMGX/build/libamgxsh.so : amgx::handle_signals(int)+0xbb
 /lib64/libpthread.so.0 : ()+0xf680
 /home/atrikut/local/AMGX/build/libamgxsh.so : amgx::Solver<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >::solve(amgx::Vector<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >&, amgx::Vector<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >&, bool)+0x11
 /home/atrikut/local/AMGX/build/libamgxsh.so : amgx::GMRES_Solver<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >::solve_iteration(amgx::Vector<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >&, amgx::Vector<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >&, bool)+0x5d2
 /home/atrikut/local/AMGX/build/libamgxsh.so : amgx::Solver<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >::solve(amgx::Vector<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >&, amgx::Vector<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >&, bool)+0x554
 /home/atrikut/local/AMGX/build/libamgxsh.so : amgx::Solver<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >::solve_no_throw(amgx::Vector<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >&, amgx::Vector<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >&, amgx::AMGX_STATUS&, bool)+0x77
 /home/atrikut/local/AMGX/build/libamgxsh.so : amgx::AMG_Solver<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >::solve(amgx::Vector<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >&, amgx::Vector<amgx::TemplateConfig<(AMGX_MemorySpace)1, (AMGX_VecPrecision)0, (AMGX_MatPrecision)0, (AMGX_IndPrecision)2> >&, amgx::AMGX_STATUS&, bool)+0x3d
 /home/atrikut/local/AMGX/build/libamgxsh.so : amgx::AMGX_ERROR amgx::(anonymous namespace)::solve_with<(AMGX_Mode)8193, amgx::AMG_Solver, amgx::Vector>(AMGX_solver_handle_struct*, AMGX_vector_handle_struct*, AMGX_vector_handle_struct*, amgx::Resources*, bool)+0x268
 /home/atrikut/local/AMGX/build/libamgxsh.so : AMGX_solver_solve()+0x147
 ./amgx_capi : main()+0x34d
 /lib64/libc.so.6 : __libc_start_main()+0xf5
 ./amgx_capi() [0x402233]
Segmentation fault (core dumped)

```

But when preconditioner is not `NOSOLVER`:

```
[atrikut@node1144 examples]$ cat ../configs/core/GMRES.json
{
    "config_version": 2,
    "solver": {
        "preconditioner": {
            "scope": "amg",
            "solver": "BLOCK_JACOBI"
        },
        "use_scalar_norm": 1,
        "solver": "GMRES",
        "print_solve_stats": 1,
        "obtain_timings": 1,
        "monitor_residual": 1,
        "convergence": "RELATIVE_INI_CORE",
        "scope": "main",
        "tolerance": 1e-6,
        "norm": "L2"
    }
}
[atrikut@node1144 examples]$ ./amgx_capi -m matrix.mtx -c ../configs/core/GMRES.json
AMGX version 2.0.0.130-opensource
Built on May 17 2018, 05:29:30
Compiled with CUDA Runtime 8.0, using CUDA driver 9.0
Warning: No mode specified, using dDDI by default.
Reading data...
RHS vector was not found. Using RHS b=[1,…,1]^T
Solution vector was not found. Setting initial solution to x=[0,…,0]^T
Finished reading
           iter      Mem Usage (GB)       residual           rate
         --------------------------------------------------------------
            Ini                   0   3.464102e+00
              0                   0   1.934535e+00         0.5585
              1              0.0000   1.934535e+00         1.0000
              2              0.0000   1.934535e+00         1.0000
              3              0.0000   1.644110e+00         0.8499
              4              0.0000   1.507196e+00         0.9167
              5              0.0000   1.213641e+00         0.8052
              6              0.0000   1.160986e+00         0.9566
              7              0.0000   1.092385e+00         0.9409
              8              0.0000   1.088166e+00         0.9961
              9              0.0000   8.810365e-01         0.8097
             10              0.0000   4.990786e-01         0.5665
             11              0.0000   4.764949e-01         0.9547
             12              0.0000   1.059929e-01         0.2224
             13              0.0000   1.677930e-15         0.0000
         --------------------------------------------------------------
         Total Iterations: 14
         Avg Convergence Rate: 		         0.0806
         Final Residual: 		   1.677930e-15
         Total Reduction in Residual: 	   4.843766e-16
         Maximum Memory Usage: 		          0.000 GB
         --------------------------------------------------------------
Total Time: 0.0461597
    setup: 0.000248704 s
    solve: 0.045911 s
    solve(per iteration): 0.00327936 s
[atrikut@node1144 examples]$

```

**Comment by marsaev** [comment]
Was able to reproduce, will look into that.

**Comment by toothstone** [comment]
I can confirm that as of today this is still an issue.

**Comment by marsaev** [comment]
Sorry for taking too long fixing the issue.
Error happens when gmres need a restart and unconditional call to preconditioner happens. In the case of `preconditioner=NOSOLVER` preconditioner is null and amgx segfaults. I'll mention the issue in the fix.

Thanks,

**Comment by marsaev** [comment]
@toothstone Can you try fix in https://github.com/NVIDIA/AMGX/commit/ac31f64cf8e4b87f0df335fd4c8cd43f0aa43a33 ?

Thanks,

**Comment by toothstone** [comment]
Hi @marsaev, thanks for the patch.

I've tried both branch v2.1.x and cherry-picking the commit to master, but although the patch fixes the issue described here (GMRES works with NOSOLVER preconditioner), I can't use GMRES **with** preconditioning now, problems which were solved flawlessly before now won't converge.

I haven't really looked at the code, but maybe the `no_preconditioner` flag is set even if there is one?
